docs/design_notes/jax_versioning.md

```diff
@@ -0,0 +1,188 @@
+The JAX-specific pieces inside XLA are primarily in the
+[`xla/python` subdirectory](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/xla/python).
+
+
+The reason that C++ pieces of JAX, such as Python bindings and runtime
+components, are inside the XLA tree is partially
+historical and partially technical.
+
+The historical reason is that originally the
+`xla/python` bindings were envisaged as general purpose Python bindings that
+might be shared with other frameworks. In practice this is increasingly less
+true, and `xla/python` incorporates a number of JAX-specific pieces and is
+likely to incorporate more. So it is probably best to simply think of
+`xla/python` as part of JAX.
+
+The technical reason is that the XLA C++ API is not stable. By keeping the
+XLA:Python bindings in the XLA tree, their C++ implementation can be updated
+atomically with the C++ API of XLA. It is easier to maintain backward and forward
+compatibility of Python APIs than C++ ones, so `xla/python` exposes Python APIs
+and is responsible for maintaining backward compatibility at the Python
+level.
+
+`jaxlib` is built using Bazel out of the `jax` repository. The pieces of
+`jaxlib` from the XLA repository are incorporated into the build
+[as a Bazel submodule](https://github.com/google/jax/blob/main/WORKSPACE).
+To update the version of XLA used during the build, one must update the pinned
+version in the Bazel `WORKSPACE`. This is done manually on an
+as-needed basis, but can be overriden on a build-by-build basis.
+
+
+## How do we make changes across the `jax` and `jaxlib` boundary between releases?
+
+The jaxlib version is a coarse instrument: it only lets us reason about
+*releases*.
+
+However, since the `jax` and `jaxlib` code is split across repositories that
+cannot be updated atomically in a single change, we need to manage compatibility
+at a finer granularity than our release cycle. To manage fine-grained
+compatibility, we have additional versioning that is independent of the `jaxlib`
+release version numbers.
+
+We maintain an additional version number (`_version`) in
+[`xla_client.py` in the XLA repository](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/xla_client.py).
+The idea is that this version number, is defined in `xla/python`
+together with the C++ parts of JAX, is also accessible to JAX Python as
+`jax._src.lib.xla_extension_version`, and must
+be incremented every time that a change is made to the XLA/Python code that has
+backwards compatibility implications for `jax`. The JAX Python code can then use
+this version number to maintain backwards compatibility, e.g.:
+
+```
+# 123 is the new version number for _version in xla_client.py
+if jax._src.lib.xla_extension_version >= 123:
+  # Use new code path
+  ...
+else:
+  # Use old code path.
+```
+
+Note that this version number is in *addition* to the constraints on the
```

Comment:
Does this doc deserve a section on how we plan to release CUDA jaxlib's too?

I want that documented so we don't keep changing it every now and then. If not here, I can document it somewhere else.

Reply:
Yeah, sounds good. I'll do it after you submit this.